program-runtime/src/instruction_processor.rs

```diff
@@ -151,13 +151,14 @@ impl PreAccount {
             }
         }
 
-        // Only the system program can change the size of the data
-        //  and only if the system program owns the account
+        // Account data size cannot exceed a maxumum length
+        if post.data().len() > MAX_PERMITTED_DATA_LENGTH as usize {
+            return Err(InstructionError::InvalidRealloc);
+        }
+
+        // The owner of the account can change the size of the data
         let data_len_changed = pre.data().len() != post.data().len();
-        if data_len_changed
-            && (!system_program::check_id(program_id) // line coverage used to get branch coverage
-                || !system_program::check_id(pre.owner()))
-        {
+        if data_len_changed && program_id != pre.owner() {
```

Comment:
All tests currently pass, can you lay out exactly what scenario you are thinking of and I'll check that we have a test for it or write one up to verify.

Reply:
I looked at this for a bit and couldn't convince myself that this would still work:
https://github.com/solana-labs/solana/blob/234461f779b1ac67c467f2d860ce4d4dfb83c720/runtime/src/system_instruction_processor.rs#L172
I'm hopefully just missing something, or we have a pretty significant test coverage hole.  Probably the former!

Reply:
My intuition around it is that when an account is passed to the program that owns it, that program has full ability to change the data len and the owner of the account at the same time. For system create account, a system account is passed to the system program which then changes data len and the owner to something new. The above check allows the data len change because the "pre" (original) owner matches the system program's id. It's not checking the new owner, that would be `post.owner()`.

Reply:
Might also be helpful to compare this new logic to the previous behaviour rewritten like this:
```suggestion
        if data_len_changed && (program_id != pre.owner()
                || !system_program::check_id(program_id)) {
```